node_modules/oae-principals/lib/api.group.js

```diff
@@ -832,6 +844,119 @@ var updateGroup = module.exports.updateGroup = function(ctx, groupId, profileFie
+    validator.check(groupId, {'code': 400, 'msg': 'A valid group id must be provided'}).isGroupId();
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    }
+
+    // Check if the user has permission to delete the group
+    canDeleteGroup(ctx, groupId, function(err, canDelete, group) {
+        if (err) {
+            return callback(err);
+        } else if (!canDelete) {
+            return callback({'code': 401, 'msg': 'You are not authorized to delete this group'});
+        }
+
+        // Mark the group as deleted
+        PrincipalsDAO.deletePrincipal(groupId, function(err) {
+            if (err) {
+                return callback(err);
+            }
+
+            // Notify consumers that a group has been deleted
+            PrincipalsEmitter.emit(PrincipalsConstants.events.DELETED_GROUP, ctx, group);
+            return callback();
+        });
+    });
+};
+
+/**
+ * Restore a group
+ *
+ * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
+ * @param  {String}     groupId         The id of the group to restore
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Obect}      callback.err    An error that occured, if any
+ */
+var restoreGroup = module.exports.restoreGroup = function(ctx, groupId, callback) {
+    var validator = new Validator();
+    validator.check(groupId, {'code': 400, 'msg': 'A valid group id must be provided'}).isGroupId();
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    }
+
+    canRestoreGroup(ctx, groupId, function(err, canRestore, group) {
+        if (err) {
+            return callback(err);
+        } else if (!canRestore) {
+            return callback({'code': 401, 'msg': 'You are not authorized to restore this group'});
+        }
+
+        // Unmark the group as deleted
+        PrincipalsDAO.restorePrincipal(groupId, function(err) {
+            if (err) {
+                return callback(err);
+            }
+
+            // Notify consumers that a group has been restored
+            PrincipalsEmitter.emit(PrincipalsConstants.events.RESTORED_GROUP, ctx, group);
+            return callback();
+        });
+    });
+};
+
+/**
+ * Determine whether or not the user in context can delete a given group
+ *
+ * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
+ * @param  {String}     groupId             The group to check
+ * @param  {Function}   callback            Standard callback function
+ * @param  {Object}     callback.err        An error that occurred, if any
+ * @param  {Boolean}    callback.canDelete  Whether or not the user can delete the group
+ * @param  {Group}      callback.group      The group object that was fetched from the database
+ */
+var canDeleteGroup = module.exports.canDeleteGroup = function(ctx, groupId, callback) {
```

Comment:
I realize it might be best to leave this to a later refactor, but I like the pattern of having these `canAction` functions in a separate `authz.js` file.